src/bin/cargo/commands/remove.rs

```diff
@@ -133,49 +137,200 @@ fn parse_section(args: &ArgMatches) -> DepTable {
     table
 }
 
-/// Clean up workspace dependencies which no longer have a reference to them.
-fn gc_workspace(workspace: &Workspace<'_>, dependencies: &[String]) -> CargoResult<()> {
+/// Clean up the workspace.dependencies, profile, patch, and replace sections of the root manifest
+/// by removing dependencies which no longer have a reference to them.
+fn gc_workspace(workspace: &Workspace<'_>) -> CargoResult<()> {
     let mut manifest: toml_edit::Document =
         cargo_util::paths::read(workspace.root_manifest())?.parse()?;
+    let mut is_modified = true;
 
     let members = workspace
         .members()
         .map(|p| LocalManifest::try_new(p.manifest_path()))
         .collect::<CargoResult<Vec<_>>>()?;
 
-    for dep in dependencies {
-        if !dep_in_workspace(dep, &members) {
-            remove_workspace_dep(dep, &mut manifest);
+    let mut dependencies = members
+        .iter()
+        .flat_map(|manifest| {
+            manifest.get_sections().into_iter().flat_map(|(_, table)| {
+                table
+                    .as_table_like()
+                    .unwrap()
+                    .iter()
+                    .map(|(key, item)| Dependency::from_toml(&manifest.path, key, item))
+                    .collect::<Vec<_>>()
+            })
+        })
+        .collect::<CargoResult<Vec<_>>>()?;
+
+    // Clean up the workspace.dependencies section and replace instances of
+    // workspace dependencies with their definitions
+    if let Some(toml_edit::Item::Table(deps_table)) = manifest
+        .get_mut("workspace")
+        .and_then(|t| t.get_mut("dependencies"))
+    {
+        deps_table.set_implicit(true);
+        for (key, item) in deps_table.iter_mut() {
+            let ws_dep = Dependency::from_toml(&workspace.root(), key.get(), item)?;
+
+            let mut is_used = false;
+            for dep in dependencies.iter_mut().filter(|d| {
+                d.toml_key() == key.get() && matches!(d.source(), Some(Source::Workspace(_)))
+            }) {
+                is_used = true;
+                *dep = ws_dep.clone();
```

Comment:
> I also noticed that workspace dependencies might not be properly handled by the likes of spec_has_match/source_has_match, so I modified the workspace.dependencies pass to replace each instance of these with the corresponding definition.

This mutation lacks enough context to need a comment

Reply:
Comment added